## Question 1 :
Classify the standard “Iris Dataset” using the decision tree algorithm. Print the 
decision tree and all the standard evaluation measures for all the classes

## Importing libraries and in-built functions 

In [22]:
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
from sklearn.tree import export_text

In [2]:
# Load the Iris dataset
iris = datasets.load_iris()
data = pd.DataFrame(data=iris.data, columns=iris.feature_names)
target = pd.Series(data=iris.target, name=iris.target_names[0])

In [3]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=1)

In [4]:
# Create a decision tree classifier
clf = DecisionTreeClassifier(random_state=1)
clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=1)

In [5]:
# Make predictions on the test set
y_pred = clf.predict(X_test)

In [6]:
# Print the decision tree
tree_rules = export_text(clf, feature_names=iris.feature_names)
print("Decision Tree Rules:")
print(tree_rules)

Decision Tree Rules:
|--- petal width (cm) <= 0.80
|   |--- class: 0
|--- petal width (cm) >  0.80
|   |--- petal width (cm) <= 1.65
|   |   |--- petal length (cm) <= 4.95
|   |   |   |--- class: 1
|   |   |--- petal length (cm) >  4.95
|   |   |   |--- sepal length (cm) <= 6.05
|   |   |   |   |--- petal width (cm) <= 1.55
|   |   |   |   |   |--- class: 2
|   |   |   |   |--- petal width (cm) >  1.55
|   |   |   |   |   |--- class: 1
|   |   |   |--- sepal length (cm) >  6.05
|   |   |   |   |--- class: 2
|   |--- petal width (cm) >  1.65
|   |   |--- petal length (cm) <= 4.85
|   |   |   |--- sepal width (cm) <= 3.10
|   |   |   |   |--- class: 2
|   |   |   |--- sepal width (cm) >  3.10
|   |   |   |   |--- class: 1
|   |   |--- petal length (cm) >  4.85
|   |   |   |--- class: 2



## Evaluation Measures

In [23]:
print("\nClassification Report:")
accuracy_dt = metrics.accuracy_score(y_test, y_pred)
precision = metrics.precision_score(y_test, y_pred, average=None)
recall = metrics.recall_score(y_test, y_pred, average=None)
f1_score = metrics.f1_score(y_test, y_pred, average=None)

print(f"Overall Accuracy: {accuracy_dt}")

print("\nEvaluation Measures for All Classes:")
for i in range(len(iris.target_names)):
    print(f"Class {i} ({iris.target_names[i]}):")
    print(f"  Precision: {precision[i]}")
    print(f"  Recall: {recall[i]}")
    print(f"  F1 Score: {f1_score[i]}")


Classification Report:
Overall Accuracy: 1.0

Evaluation Measures for All Classes:
Class 0 (setosa):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0
Class 1 (versicolor):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0
Class 2 (virginica):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0


In [8]:
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Confusion Matrix:
[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]


In [9]:
print("\nAccuracy Score:", accuracy_score(y_test, y_pred))


Accuracy Score: 0.9666666666666667


## Question 2

Classify the same dataset using Logisitic Regression. Perform 5-fold cross 
validation and print the overall accuracy for all the folds. For the best model, print 
the confusion matrix and all the standard evaluation measures for all the classes.

In [10]:
# Import necessary libraries
from sklearn import datasets
from sklearn.model_selection import cross_val_predict, cross_val_score, train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report

In [11]:
# Load the Iris dataset
iris = datasets.load_iris()
X = iris.data
y = iris.target

In [12]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [26]:
# Create a logistic regression classifier
LRclassifier = LogisticRegression()

In [28]:
# Perform 5-fold cross-validation
cv_scores = cross_val_score(LRclassifier, X_train, y_train, cv=5)

C:\Users\anant\AppData\Local\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\anant\AppData\Local\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#l

In [29]:
# Print the accuracy for each fold
print("Cross-Validation Accuracies:")
for i, accracy in enumerate(cv_scores, 1):
    print(f"Fold {i}: {accracy}")

Cross-Validation Accuracies:
Fold 1: 1.0
Fold 2: 1.0
Fold 3: 0.875
Fold 4: 1.0
Fold 5: 0.9583333333333334


In [30]:
# Fit the classifier on the training data
LRclassifier.fit(X_train, y_train)

LogisticRegression()

In [31]:
y_pred = LRclassifier.predict(X_test)

In [33]:
accuracy_lr = metrics.accuracy_score(y_test, y_pred)
precision = metrics.precision_score(y_test, y_pred, average=None)
recall = metrics.recall_score(y_test, y_pred, average=None)
f1_score = metrics.f1_score(y_test, y_pred, average=None)



print("\nEvaluation Measures for All Classes:")
for i in range(len(iris.target_names)):
    print(f"Class {i} ({iris.target_names[i]}):")
    print(f"  Precision: {precision[i]}")
    print(f"  Recall: {recall[i]}")
    print(f"  F1 Score: {f1_score[i]}")


Evaluation Measures for All Classes:
Class 0 (setosa):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0
Class 1 (versicolor):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0
Class 2 (virginica):
  Precision: 1.0
  Recall: 1.0
  F1 Score: 1.0


In [18]:
# Print the confusion matrix
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Confusion Matrix:
[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


In [19]:
# Print evaluation measures
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



# Question 3:
## Compare the results obtained using Decision tree and Logistic Regression. Write your inference on the 5-fold cross validation results obtained for Q.2

## Dataset Description
The Iris dataset is used for this assignment, which consists of 3 different types of irises’ (Setosa, Versicolour, and Virginica) petal and sepal length, stored in a 150x4 numpy.ndarray

The rows being the samples and the columns being: Sepal Length, Sepal Width, Petal Length and Petal Width.
## Comparision of models
The Iris dataset is used for evaluating the Decesion Tree Algorithm and Logistic Regression, with data split of 80% of Training and 20% of Testing for both models , where  Logistic Regression outperforms the Decision tree since the overall accuracy of Logistic Regression is 97.3% and for decision tree it is 96%.

Logistic Regression has better precision, recall, and F1 scores than Decision tree for every class.


Based on the cross fold accuracies we can infer the following:

The model shows a high level of consistency across all five folds. Accuracy is consistently high, indicating that the model generalizes well to different subsets of the data. The average accuracy across all folds is quite high, with the lowest accuracy being 87.5% and the highest being 100%.

Folds 1,2 and 4 have  achieved perfect accuracy (100%), indicating that the model performed exceptionally well on those subsets of the data. 
Folds 3 and 5 also exhibit high accuracy, further supporting the overall effectiveness of the model.

This suggests that the model is effective in correctly classifying instances in the dataset.

Decision trees tend to capture complex relationships in the data, potentially leading to overfitting. Therefore, the accuracy on the training set might be high, but it might not generalize well to new, unseen data.
Logistic regression is a linear model that is less prone to overfitting, especially when dealing with a relatively small dataset like the Iris dataset. It provides a simpler decision boundary.